src/Orleans.Analyzers/NoRefParamsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,55 @@
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Orleans.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class NoRefParamsDiagnosticAnalyzer : DiagnosticAnalyzer
+    {
+        private const string BaseInterfaceName = "IGrain";
+        public const string DiagnosticId = "ORLEANS0002";
+        public const string Title = "Reference parameter modifiers are not allowed";
+        public const string MessageFormat = Title;
+        public const string Category = "Usage";
+
+        private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(AnalyzeSyntax, SyntaxKind.MethodDeclaration);
+        }
+
+        private static void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            if (!(context.Node is MethodDeclarationSyntax syntax)) return;
+
+            var symbol = context.SemanticModel.GetDeclaredSymbol(syntax);
+            var isClassOrInterface = symbol.ContainingType.TypeKind == TypeKind.Interface ||
+                                     symbol.ContainingType.TypeKind == TypeKind.Class;
```

Comment:
We should only do this for interface, I think - any objections?

Reply:
Checking only for interfaces seems better, because to write a Grain you would always need the interface, and the only methods that should be checked would be the ones delcared in the interface, is that right?

Reply:
That's right